### ISSUE TITLE
Improve the memory handling for the model optimization

### ISSUE THREAD

**Issue by tobias-liaudat** [CosmoStat/wf-psf#4]
Right now the notebook psf_field_v2.ipynb contains the routines that can optimise the (_simplest_) PSF model.
This is a complete parametric approach but it is good to have a benchmark and to evaluate the performance of a fully parametric model.

Bear in mind that the data generated, the train and test PSF fields, were done with the same parametric model (_Inverse crime!_). 

Right now what we do is:

1. Start with a bunch of positions and observed stars. ie 100 or 200 to start.
2. Try to estimate the Zernike coefficients for each observed star by using the input positions and the model parameters.
3. For that we calculate all the Zernike coefficients and the OPD in a vectorial way.
4. Then, we have to calculate the polychromatic PSF in a serial manner for each star (using a while_loop).
5. Finally, each polychromatic calculation goes in a serial manner as we calculate each monochromatic PSF before doing the weighted average with the calculated SED values.

This strategy is consuming all the GPU memory when we have to optimise. Not when we calculate the loss (forward pass) but when we declare the trainable model variables and ask tf to optimise them.

We need to reduce the memory in the optimisation. Approaches:

- [x] Reduce dimensions of OPD, phase matrices. (To see if there is aliasing with the fine spider obscurations)
- [x] Choose intelligently the data-types of the different tensors. See if it's possible to do the fft on other than tf.complex128 data type.
- [x] Perform batch optimisation! It is too heavy to calculate the back-prop over all the input dataset :o Mutate into a more ML framework and use batch optimisation with a low batch number.

Good luck!

**Comment by tobias-liaudat** [comment]
The new parametric PSF model was implemented with a more Tensorflow spice, inheriting from the Keras.Layer and Keras.Model.

Now the model can be compiled! It does not run in eager mode anymore. The problem was with the difference between:
`tf.shape(Tensor)` and `Tensor.shape`. When compiled, the first one is evaluated in runtime while the second is evaluated at compilation time (we don't know the proper size of the padding needed for a specific wavelength bin during the compilation time). 

Right now it is possible to optimise the model (on Colab) without a memory crash and it is almost twice as faster than before!

The Colab notebook is here:  #https://github.com/tobias-liaudat/wf-psf/commit/7d57df7db41f3a115780b0638aca0fd1b6e53fd5 

The tensorflow code is here: #https://github.com/tobias-liaudat/wf-psf/commit/7d5653c04b760c7c3fb8cc87a67d36cb2d60d4aa
